### PR TITLE
feat: support k8s db ssl [DET-6364]

### PR DESCRIPTION
**Pull request by eecsliu**
## Description
Lets users specify a secret name and ssl certificate name for postgres. Ssl mode is configurable, but only `verify-ca` is supported at this moment.

## Test Plan
Uncomment the relevant fields in `db` of `values.yaml` and deploy using helm. You might have to generate keys/certificates and create a secret in kubernetes first. Then, ssh into the master container and check to see that the secret you specified is present under `/mount/determined/secrets/` and that the `ssl_root_cert` is populated in the master config.

Additionally, you may deploy a db instance with ssl enabled. Add your certificate to the `db` fields and check that your master instance is able to launch and connect to the db. Steps here:

Start by generating keys and create a kubernetes secret.
```
openssl req -new -x509 -days 365 -nodes -text -out server.crt \
  -keyout server.key -subj "/CN=dbhost.fakedomain.com"

chmod og-rwx server.key

openssl req -new -nodes -text -out root.csr \
  -keyout root.key -subj "/CN=root.fakedomain.com"
chmod og-rwx root.key

openssl x509 -req -in root.csr -text -days 3650 \
  -extfile /etc/ssl/openssl.cnf -extensions v3_ca \
  -signkey root.key -out root.crt

openssl req -new -nodes -text -out server.csr \
  -keyout server.key -subj "/CN=dbhost.fakedomain.com"
chmod og-rwx server.key

openssl x509 -req -in server.csr -text -days 365 \
  -CA root.crt -CAkey root.key -CAcreateserial \
  -out server.crt

kubectl create secret tls tls-secret-postgres --cert=server.crt --key=server.key
```
Create a file named `db.yaml` and insert the following:
```
tls:
  enabled: true
  certificatesSecret: tls-secret-postgres
  certFilename: tls.crt
  certKeyFilename: tls.key
volumePermissions:
  enabled: true
```

Install bitnami postgres with the config you just create:
```
helm install postgresql bitnami/postgresql -f db.yaml
```

Find the password for the postgres user and login:
```
export PASS=$(kubectl get secret postgresql -o jsonpath="{.data.postgresql-password}" | base64 --decode) 
kubectl exec -it pod/postgresql-0 -- psql -U postgres
```

Once logged in, create a user `determined` with password `determined`. Also create a new table named `determined`. Verify that the user has been created and can be logged into.

Modify the `values.yaml` file in the helm chart so that the db uses `determined` as username and password.  Then, create a configmap with the certificate:
```
kubectl create cm root-cert -n determined --from-file=certs/root.crt
```

Use `kubernetes get services` to identify the name of the postgres service. Copy the ip address of the database and edit `master-deployment.yaml`. We want to add the `hostAliasses` field at the same level as the `containers` field:
```
containers:
- image: determinedai/determined-master:0.17.9-rc1
  imagePullPolicy: Always
  name: determined-master-determined-ai
  .
  .
hostAliases:
- hostnames:
  - dbhost.fakedomain.com
  ip: <postgres IP address>
```

Finally, populate the `values.yaml` DB fields for `ssl_mode`:
```
sslMode: verify-ca
sslRootCert: root.crt
certSecretName: root-cert
resourceType: configmap
```

run `helm install <deployment name> <local path to determined/helm/charts/determined>` and check the log of the master instance with `kubectl logs <master instance ID>`. Make sure the logs indicate the instance is running properly; any DB issue should result in it not being able to connect and Determined not launching.

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by gh-determined-ai** [comment]
Wouldn't want you to be the first to have to do this but a new release note format is outlined here: 
[](https://determinedai.atlassian.net/wiki/spaces/DOC/pages/1101594663/Documentation+Process#Release-Note-Template)
Especially for new features, there should be more description of the benefits.

**Comment by eecsliu** [comment]
> Wouldn't want you to be the first to have to do this but a new release note format is outlined here: Especially for new features, there should be more description of the benefits. Your feedback on the release note documentation process will be helpful.

Ah I forgot - thanks for the reminder!

File: helm/charts/determined/templates/master-config.yaml [hunk 1]
```diff
@@ -45,6 +45,11 @@ data:
       {{- end  }}
       port: {{ .Values.db.port }}
       name: {{ .Values.db.name | quote }}
+      {{- if .Values.db.sslMode }}
+      ssl_mode: {{ .Values.db.sslMode }}
+      {{- $rootCert := (required "A valid .Values.db.sslRootCert entry required!" .Values.db.sslRootCert )}}
```
Comment (ioga) [inline helm/charts/determined/templates/master-config.yaml:50]: when using `sslMode`, are `certResourceName` and `resourceType` also required? should we add `required` statement for them as well?

or is there a way to inject and use the cert without these.

**Review by ioga — APPROVED** [review]

File: helm/charts/determined/templates/master-config.yaml (hunk 1 above)
Comment (eecsliu) [reply helm/charts/determined/templates/master-config.yaml:50]: yeah, I think we should already be checking for those in `master-deployment.yaml`

**Pull request merged**